test/files/pos/InferOverloadedPartialFunction.scala

```diff
@@ -0,0 +1,41 @@
+class MySeq[T] {
+  def map1[U](f: T => U): MySeq[U] = new MySeq[U]
+  def map2[U](f: T => U): MySeq[U] = new MySeq[U]
+}
+
+class MyMap[A, B] extends MySeq[(A, B)] {
+  def map1[C](f: (A, B) => C): MySeq[C] = new MySeq[C]
+  def map1[C, D](f: (A, B) => (C, D)): MyMap[C, D] = new MyMap[C, D]
+  def map1[C, D](f: ((A, B)) => (C, D)): MyMap[C, D] = new MyMap[C, D]
+
+  def foo(f: Function2[Int, Int, Int]): Unit = ()
+  def foo[R](pf: PartialFunction[(A, B), R]): MySeq[R] = new MySeq[R]
+}
+
+object Test {
+  val m = new MyMap[Int, String]
+
+  // This one already worked because it is not overloaded:
+  m.map2 { case (k, v) => k - 1 }
+
+  // These already worked because preSelectOverloaded eliminated the non-applicable overload:
+  m.map1(t => t._1)
+  m.map1((kInFunction, vInFunction) => kInFunction - 1)
+  val r1 = m.map1(t => (t._1, 42.0))
+  val r1t: MyMap[Int, Double] = r1
+
+  // These worked because the argument types are known for overload resolution:
+  m.map1({ case (k, v) => k - 1 }: PartialFunction[(Int, String), Int])
+  m.map2({ case (k, v) => k - 1 }: PartialFunction[(Int, String), Int])
+
+  // These ones did not work before:
+  m.map1 { case (k, v) => k }
+  val r = m.map1 { case (k, v) => (k, k*10) }
+  val rt: MyMap[Int, Int] = r
+  m.foo { case (k, v) => k - 1 }
```

Comment:
It's not quite unrelated because getting this part right is what led to the 2nd change. Look the definitions of `map1` in the latest version:

```scala
class MySeq[T] {
  def map1[U](f: T => U): MySeq[U] = new MySeq[U] // 1
}

class MyMap[A, B] extends MySeq[(A, B)] {
  def map1[C](f: (A, B) => C): MySeq[C] = new MySeq[C] // 2
  def map1[C, D](f: (A, B) => (C, D)): MyMap[C, D] = new MyMap[C, D] // 3
  def map1[C, D](f: ((A, B)) => (C, D)): MyMap[C, D] = new MyMap[C, D] // 4
```

1 is the original `Function1`-based version, 2 and 3 are the `Function2`-based versions that we want for `MyMap`. The only reason for having 4 in addition to 3 (which should be more efficient because it doesn't have to tuple the arguments) is to support pattern-matching anonymous functions (for which `Function2` is no longer applicable in overload resolution), otherwise you would always get 1 (with the wrong return type) in this case.